torch/_functorch/aot_autograd.py

```diff
@@ -1899,6 +1899,43 @@ def functionalized_f_helper(*args):
             # Run the joint
             f_outs = fn(*f_args)
 
+        if trace_joint:
+            # We support a limited amount of mutation of graph inputs during the backward pass.
+            # (This is used e.g. by Float8, which needs to update buffers during the backward pass)
+            # Here, we perform extra checks for primals that were mutated in the **backward**
+            # We're doing the checks here instead of doing them with the rest of the input mutation handling because:
+            # - We need to detect inputs that were mutated in the backward **separately** from mutations that happened
+            #   during the forward, because the handling is different: some input mutations from the the forward
+            #   can be only handled in a fw-only runtime epilogue, and in theory if we wanted to handle those same
+            #   types of mutations in the backward we would need a bw-only runtime epilogue.
+            # - We could in theory have our analysis pass differentiate mutations in the fw from mutations in
+            #   the bw by running our analysis first on the fw-only graph, and then on the joint graph. This would
+            #   require an extra round of tracing though, so it's more efficient to do in-line here.
+            assert isinstance(args, tuple) and len(args) == 2 and isinstance(args[0], (list, tuple))
+            # Only look at mutations that happened to forward inputs (e.g. fw buffers that were saved for bw)
+            primals_before = args[0]
+            primals_after = pytree.tree_map(from_fun, f_args[0])
+            for before, after, inpt_info in zip(primals_before, primals_after, meta.input_info):
+                # Ban metadata mutations on fw inputs during the bw
+                if not inpt_info.mutates_metadata:
+                    assert not was_metadata_updated(before, after), \
+                        "Found a graph input that had its metadata mutated in the backward. This is not supported"
```

Comment:
Curious question what is an example of metadata that could be mutated in the backward?

Reply:
Very late reply - the only example I can think of is if you have a (horrible) custom `autograd.Function`, that directly mutates one of its `grad_out` arguments in the backward. e.g.:

```
class FuncIHopeNobodyWrites(torch.autograd.Function):
    ...
    def backward(ctx, grad_x):
        grad_x.t_(1, 0)
        return grad_x * 2
        
x = torch.ones(2, 4, requires_grad=True)
out = FuncIHopeNobodyWrites.apply(x)

gradient = torch.ones(2, 4)
# the custom func will mutate gradient directly!
out.backward(gradient)
```